src/evaluators/BinaryExpression.js

```diff
@@ -194,11 +194,28 @@ export function computeBinary(
 
   let resultType;
   const compute = () => {
-    if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
+    let lvalIsAbstract = lval instanceof AbstractValue;
+    let rvalIsAbstract = rval instanceof AbstractValue;
+
+    if (lvalIsAbstract || rvalIsAbstract) {
       try {
         // generate error if binary operation might throw or have side effects
         resultType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-        return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
+
+        // We have determined that it is safe to assume the type of this operation,
+        // but there is one last check. If the operation can throw, then our assumption
+        // may rely on a dynamic condition. We express this condition as an implication,
+        // rval is temporal => this operation is temporal
+        // See #2327
+        if ((op === "in" || op === "instanceof") && (rvalIsAbstract && rval.isTemporal()))
+          return AbstractValue.createTemporalFromBuildFunction(
+            realm,
+            resultType,
+            [lval, rval],
+            createOperationDescriptor("BINARY_EXPRESSION", { op }),
+            { isPure: true, skipInvariant: true }
+          );
+        else return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
```

Comment:
I'm still a little bit disturbed by this. If one of the arguments is temporal, then moving the operation to point where the current path conditions do not apply may be safe, but it seems hardly desirable.

In essence, any operation over temporal values should result in a temporal value, as a rule, except for a few specific cases that are well documented.

Reply:
I'm going to enforce this protocol in a follow up PR that addresses #2489. Doing so will involve auditing uses of all of the abstract value factory methods.